### PR TITLE
ovn-k8s-cni-overlay: implement bandwidth shaping outside OVN

### PR DESCRIPTION
**Pull request by dcbw**
OVN's QoS needs some clarification, as the manpages state that the QoS table only works with userspace DPDK.  Until that happens (and if true, QoS gets implemented for non-DPDK), implement bandwidth shaping by porting the openshift-sdn code over that handles it directly in OVS.

Not supported on Windows for now; @alinbalutoiu any idea what that would take?

@rajatchopra @shettyg

### CODE REVIEW

**Comment by alinbalutoiu** [comment]
CC @aserdean

**Comment by dcbw** [comment]
@alinbalutoiu @aserdean I actually took a stab at the Windows side, see if that looks correct.

File: go-controller/cmd/ovn-k8s-cni-overlay/app/helper_linux.go [hunk 1]
```diff
@@ -103,8 +103,69 @@ func setupInterface(netns ns.NetNS, containerID, ifName, macAddress, ipAddress,
...
+			return err
+		}
+		err = ovsSet("port", hostVeth, fmt.Sprintf("qos=%s", qos))
+		if err != nil {
+			return err
+		}
+	}
+	if egressBPS > 0 {
+		// ingress_policing_rate is in Kbps
+		err := ovsSet("interface", hostVeth, fmt.Sprintf("ingress_policing_rate=%d", egressBPS/1024))
```
Comment (jsitnicki) [inline go-controller/cmd/ovn-k8s-cni-overlay/app/helper_linux.go]: I'm wondering if it shouldn't be `/1000`. [OVS docs for QoS](https://github.com/openvswitch/ovs/blob/master/Documentation/faq/qos.rst) give examples using 1000's as multiplier (i.e. kilo- not kibi- bps).

Comment (dcbw) [reply go-controller/cmd/ovn-k8s-cni-overlay/app/helper_linux.go]: @jsitnicki good question, I'm happy to change it

Comment (dcbw) [reply go-controller/cmd/ovn-k8s-cni-overlay/app/helper_linux.go]: @jsitnicki fixed.

**Comment by shettyg** [comment]
@dcbw 

You have already seen that this does not get erased when ovn-controller restarts?

**Comment by dcbw** [comment]
> You have already seen that this does not get erased when ovn-controller restarts?

@shettyg you mean if ovn-controller restarts for some reason while pods are active?

**Comment by aserdean** [comment]
@dcbw unfortunately QoS is not supported in the Windows datapath at the moment.

Hyper-V vSwitch which supports it for VM ports however I have no idea if it works for containers also.

I will do some testing and let you know.

Do you mind leaving configuration the same for Windows, and add some logging so that the is informed that this feature is not supported?

**Comment by shettyg** [comment]
> you mean if ovn-controller restarts for some reason while pods are active?

Yeah. I don't think there is anything in ovn-controller's code that does it though...

**Comment by shettyg** [comment]
lgtm on linux side. Waiting for Alin's feedback on windows side.

**Comment by aserdean** [comment]
Update on Windows (For Windows version > 14393):
If we would like to make a hard dependancy on Hyper-V feature (virtualization enabled), we could either use the powershell cmdlet or add the WMI provider implementation inside OVS(same as for the current linux code):
https://docs.microsoft.com/en-us/powershell/module/hyper-v/set-vmnetworkadapter?view=win10-ps 
(Argument `MinimumBandwidthAbsolute` and `MaximumBandwidth`, optional `MinimumBandwidthWeight`)
For container feature only the only thing I could find is:
https://github.com/docker/libnetwork/pull/1061/commits/b2ec240bce4b2a6bed5febdda22d726e009a8ea8 

Below is a snippet of using the cmdlets for additional information.
<details>
PS C:\> ovs-vsctl.exe show
2fb03ee8-51e3-4272-a8a1-30e3ea5454eb
    Bridge management
        Port management
            tag: 0
            Interface management
                type: internal
        Port Ethernet
            Interface Ethernet
PS C:\> Get-VMNetworkAdapter -ManagementOS | Set-VMNetworkAdapter -MaximumBandwidth 1200000000
PS C:\> C:\Users\Administrator\Desktop\iperf-3.1.3-win64\iperf3.exe -c 10.8.13.85 -i 10
Connecting to host 10.8.13.85, port 5201
[  4] local 10.8.0.156 port 50815 connected to 10.8.13.85 port 5201
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec  1.01 GBytes   867 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec  1.01 GBytes   867 Mbits/sec                  sender
[  4]   0.00-10.00  sec  1.01 GBytes   867 Mbits/sec                  receiver

iperf Done.
PS C:\> Get-VMNetworkAdapter -ManagementOS | Set-VMNetworkAdapter -MaximumBandwidth 120000000
PS C:\> C:\Users\Administrator\Desktop\iperf-3.1.3-win64\iperf3.exe -c 10.8.13.85 -i 10
Connecting to host 10.8.13.85, port 5201
[  4] local 10.8.0.156 port 50817 connected to 10.8.13.85 port 5201
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec   139 MBytes   117 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec   139 MBytes   117 Mbits/sec                  sender
[  4]   0.00-10.00  sec   139 MBytes   117 Mbits/sec                  receiver

iperf Done.
PS C:\> Get-VMNetworkAdapter -ManagementOS | Set-VMNetworkAdapter -MaximumBandwidth 12000000
PS C:\> C:\Users\Administrator\Desktop\iperf-3.1.3-win64\iperf3.exe -c 10.8.13.85 -i 10
Connecting to host 10.8.13.85, port 5201
[  4] local 10.8.0.156 port 50819 connected to 10.8.13.85 port 5201
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec  14.5 MBytes  12.2 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec  14.5 MBytes  12.2 Mbits/sec                  sender
[  4]   0.00-10.00  sec  14.5 MBytes  12.2 Mbits/sec                  receiver

iperf Done.
PS C:\>
</details>

As an initial idea(since it is easier to implement), I would suggest to have a check if Hyper-V is enabled on the host and if it is, apply QoS via cmdlets, otherwise just log it is not supported w/o Hyper-V.

We can change this in the future as further pieces may be added to HNS and other enhancements are implemented inside OVS.

File: go-controller/cmd/ovn-k8s-cni-overlay/app/helper_windows.go [hunk 2]
```diff
@@ -194,6 +194,13 @@ func ConfigureInterface(args *skel.CmdArgs, namespace string, podName string, ma
 		return nil, fmt.Errorf("failure in plugging pod interface: %v  %q", err, string(out))
 	}
 
+	if err = clearPodBandwidth(args.ContainerID); err != nil {
```
Comment (alinbalutoiu) [inline go-controller/cmd/ovn-k8s-cni-overlay/app/helper_windows.go]: I would say to comment out those functions and just leave them there. Add a TODO as well for them please, we will take care of it.

Also as @aserdean suggested, put a log message saying that QoS is not supported yet on Windows.

**Comment by dcbw** [comment]
@alinbalutoiu commented out and added a commit log message.  PTAL thanks!

File: go-controller/cmd/ovn-k8s-cni-overlay/app/ovs.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	if err != nil {
+		return "", fmt.Errorf("failed to run 'ovs-vsctl %s': %v\n  %q", strings.Join(args, " "), err, string(output))
+	}
+
+	outStr := string(output)
+	if outStr != "" {
+		// If output is a single line, strip the trailing newline
+		nl := strings.Index(outStr, "\n")
+		if nl == len(outStr)-1 {
+			outStr = outStr[:nl]
```
[the remark this answers is not in the document]
Comment (dcbw) [reply go-controller/cmd/ovn-k8s-cni-overlay/app/ovs.go]: @alinbalutoiu fixed that up in a slightly different way, PTAL, thanks!

**Comment by dcbw** [comment]
@shettyg @rajatchopra PTAL thanks!

File: go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go [hunk 4]
```diff
@@ -40,6 +41,47 @@ func argString2Map(args string) (map[string]string, error) {
...
+	if rsrc.Value() > maxRsrc.Value() {
+		return fmt.Errorf("resoruce is unreasonably large (> 1Pbit)")
+	}
+	return nil
+}
+
+func extractPodBandwidthResources(podAnnotations map[string]string) (int64, int64, error) {
+	ingress := int64(-1)
+	egress := int64(-1)
+	str, found := podAnnotations["kubernetes.io/ingress-bandwidth"]
```
Comment (rajatchopra) [inline go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go]: Is this admission controlled in any way? Not affects this PR, but curious.

**Comment by rajatchopra** [comment]
/lgtm
This will have some conflicts with #354, so hopefully and selfishly that gets merged first :).

**Pull request merged**